### PR TITLE
Worksheet C

### PR DESCRIPTION
**Pull request by AZ4321**

### CODE REVIEW

**Comment by AZ4321** [comment]
Here is the Worksheet C. Check the latest merge as I had some conflicts in the other 2, it took me 2 merges to figure that out.

File: README.md [hunk 1]
```diff
@@ -1,2 +1,70 @@
...
+	if i =/ j and list[i] = list[j] then
+	return true
+	end if
+	end for
+	end for
+	return false
+	end procedure
+
+
+A) O(n). This algorithm sets two lists with i/j and for every item in the list, it takes away 1 from the length of the list and then checking if both lists are equal and if they are then return true and if not, return false.
```
Comment (edpowley) [inline README.md:21]: This seems to be explaining the operation of the algorithm line by line -- what is the task it actually performs? It checks a list for duplicated elements.

File: README.md [hunk 2]
```diff
@@ -1,2 +1,70 @@
...
+	end if
+	end for
+	end for
+	return false
+	end procedure
+
+
+A) O(n). This algorithm sets two lists with i/j and for every item in the list, it takes away 1 from the length of the list and then checking if both lists are equal and if they are then return true and if not, return false. 
+
+B) It would be the worst case scenario because we would have go through the list n times and with each iteration, you would have to go through the first loop. Which would take double the time.
```
Comment (edpowley) [inline README.md:23]: ✔️ though strictly speaking it's not double the time, it's n times the time.

File: README.md [hunk 3]
```diff
@@ -1,2 +1,70 @@
...
+	for j = 0,..., i - 1 do
+	if i =/ j and list[i] = list[j] then
+	return true
+	end if
+	end for
+	end for
+	return false
+	end procedure
+
+c) This loop will carry out the same operation but instead of referring to the number of lists in the (j), it will link back to the first (i) list. The result of the (i) list -1 will determine (j).
```
Comment (edpowley) [inline README.md:39]: Not sure what you mean here. I was looking for something like: the first algorithm compares each pair of elements twice (i.e. A==B and B==A), whereas the second algorithm eliminates the redundant checks

File: README.md [hunk 4]
```diff
@@ -1,2 +1,70 @@
...
+
+d) The second algorithm will perform better because the function depends on the loop within the list (i) as well as the number of operations is halved as the number of elements is less. This algorithm is O(1).
```
Comment (edpowley) [inline README.md:41]: The running time still depends on n so the algorithm is not O(1)

File: README.md [hunk 5]
```diff
@@ -1,2 +1,70 @@
...
+
+e) This algorithm is no longer quadratic because it doesn't take another number for the length of the list but instead uses the first list - 1 to figure out what the value is which reduces the operation time as there is a lack of elements.
```
Comment (edpowley) [inline README.md:43]: The algorithm still has two nested loops, which still depend on n (though only indirectly for the inner loop), therefore the algorithm is still quadratic.

File: README.md [hunk 6]
```diff
@@ -1,2 +1,70 @@
...
+	if list[i - 1] = list[i] then
+	return true
+	end if
+	end for
+	return false
+	end procedure
+
+f) O(n log n)
+
+From: [https://wiki.python.org/moin/TimeComplexity](https://wiki.python.org/moin/TimeComplexity "Reference")
```
Comment (edpowley) [inline README.md:61]: ✔️

File: README.md [hunk 7]
```diff
@@ -1,2 +1,70 @@
...
+
+g) Using binary search to find largest/smallest number in the tree and the basic concept of this is to reduce the problem size with each iteration. As the number of elements increases, the operation time increases. The time complexity is bad.
```
Comment (edpowley) [inline README.md:63]: This isn't a binary search. It's a sort followed by a linear search (a single for loop for i=1 to n), so the time complexity is O(n log n + n) = O(n log n)

File: README.md [hunk 8]
```diff
@@ -1,2 +1,70 @@
...
+
+h) If the input list is large, then the O(n log n) would be the fastest as a with a large list, this sort will try break down the problem with each iteration but with O(n), it is more linear and require more steps to perform the sort.
```
Comment (edpowley) [inline README.md:65]: ✔️ though the key observation is that n log n grows more slowly than n^2 as n grows large

File: README.md [hunk 9]
```diff
@@ -1,2 +1,70 @@
...
+
+i) The slower algorithm may perform less elements but in a faster time rather than more elements in a slower time.
```
Comment (edpowley) [inline README.md:67]: Not sure what you mean here

**Pull request closed without merging**